chainer/functions/loss/softmax_cross_entropy.py

```diff
@@ -328,37 +362,41 @@ def _double_backward_softmax_cross_entropy(x, t, normalize, class_weight,
 
     if class_weight is not None:
         shape = [1 if d != 1 else -1 for d in six.moves.range(x.ndim)]
-        class_weight = chainer.functions.broadcast_to(
-            class_weight.reshape(shape), x.shape)
+        class_weight = F.broadcast_to(class_weight.reshape(shape), x.shape)
         # TODO(niboshi): Remove this workaround after ChainerX supports
         # type promotion.
         if is_chainerx:
-            class_weight = chainer.functions.cast(class_weight, x.dtype)
+            class_weight = F.cast(class_weight, x.dtype)
         loss = loss * class_weight
 
     in_use = (t != ignore_label).astype(x.dtype)
 
-    loss = chainer.functions.rollaxis(loss, 1, loss.ndim)
-    loss = chainer.functions.reshape(loss, (-1, loss.shape[-1]))
+    loss = F.rollaxis(loss, 1, loss.ndim)
+    loss = F.reshape(loss, (-1, loss.shape[-1]))
 
     # Replace ignore_label value with one valid for F.select_item below.
     t = t.clip(0, loss.shape[1] - 1)
 
-    loss = chainer.functions.select_item(loss, t.ravel())
-    loss = chainer.functions.reshape(loss, t.shape)
+    loss = F.select_item(loss, t.ravel())
+    loss = F.reshape(loss, t.shape)
 
     loss = loss * in_use
 
     if reduce == 'mean':
+        reduc_dtype = _reduction_dtype(x.dtype)
         if normalize:
-            count = in_use.sum()
+            # TODO(niboshi): Use in_use.sum(dtype=reduc_dtype) once chainerx
+            # supports dtype argument.
+            count = in_use.astype(reduc_dtype, copy=False).sum()
         else:
             count = len(x)
         count = max(count, 1.)
-        loss = loss / count
-        return chainer.functions.sum(loss)
-    else:
-        return loss
+        # Perform reduction in a promoted dtype
+        loss = F.cast(loss, reduc_dtype)
+        loss = F.sum(loss / count)
+        loss = F.cast(loss, x.dtype)
```

Comment:
How about skipping these calls to `F.cast` in case dtypes are the same?

Reply:
Fixed.